### PR TITLE
Typed static/mirror pod UID translation

### LINKED ISSUE #36031: Leverage the type system to avoid Static pod / mirror pod translation issues.
We have a lot of calls to `TranslatePodUID` through our codebase, and it can be difficult to keep track of where it's needed. We should leverage the type system with an internal type for PodUIDs that have been resolved to a static pod UID so it's clear when the translation is or is not needed.

/cc @yujuhong

### PR DESCRIPTION
**Pull request by ebbeelsborg**
Fixes #36031 , partially.

TODO:
- [x] Add types ResolvedPodUID and MirrorPodUID.
- [x] Use the ResolvedPodUID type with minimal changes.
- [x] Use the MirrorPodUID type with minimal changes.
- [x] Clarify whether the new types should be used anywhere else; if so make the agreed upon changes.

```NONE
```

### CODE REVIEW

**Comment by ebbeelsborg** [comment]
@timstclair I've now completed the basic parts. At this point I'd like to discuss how to proceed; whether to further disseminate the new types or whether to wrap this PR up. There are more places where types.UID can be specialized, but that'll affect consumers that seem to not care about distinguishing static and mirror pod UIDs. Key spots are the GetPodByUID function of pod_manager.go and the podStatuses map of the manager struct in status_manager.go. Your opinion, please? Thanks.

**Comment by ebbeelsborg** [comment]
@tallclair Just a heads up: It'd be excellent to have at least some of this work finished (merged) this week as I'll be away the rest of the month from next week (so as to avoid the PR rotting), if possible.

**Comment by ebbeelsborg** [comment]
@tallclair I'm back from vacation now and would like to finish this work. Could you please comment on my question posed above some weeks ago so that we can move forward -- or should I just go ahead and ask for a formal review of the work so far and then discuss the final point later (possibly in another issue/PR)? Thanks.

File: pkg/kubelet/status/status_manager.go [hunk 1]
```diff
@@ -155,7 +155,13 @@ func (m *manager) Start() {
 func (m *manager) GetPodStatus(uid types.UID) (v1.PodStatus, bool) {
 	m.podStatusesLock.RLock()
 	defer m.podStatusesLock.RUnlock()
-	status, ok := m.podStatuses[m.podManager.TranslatePodUID(uid)]
+	// Func removeOrphanedPodStatuses in kubelet_pods.go calls
+	// func RemoveOrphanedStatuses in status_manager.go with both
+	// static pod UIDs and mirror pod UIDs, which in turn
+	// iterates over the podStatuses map of this manager struct
+	// that is then forced to hold types.UID keys.
+	// Thus, we resolve and type cast back again.
```
Comment (tallclair) [inline pkg/kubelet/status/status_manager.go]: I don't totally understand this comment. Is this just explaining why the podStatuses map needs to key off of a types.UID type?

**Review by tallclair — COMMENTED** [review]
Welcome back. Sorry for missing your comments before you went on vacation. To answer your question, I think it would be a good idea to convert `Manager.GetPodByUID` to take a resolved ID, but if that turns into a large change I'd also take a TODO (referencing #36031). IIUC, the `podStatuses` map currently holds both static & mirror pod statuses, so I think it makes sense to keep that keyed off `types.UID`.

File: pkg/kubelet/status/status_manager.go (hunk 1 above)
Comment (ebbeelsborg) [inline pkg/kubelet/status/status_manager.go]: Yes. Let's just delete this comment?

**Comment by ebbeelsborg** [comment]
To address the question of whether/how to further disseminate the new types, I suggest to:
- Merge this PR into master (after rebase, possibly squashed, passing tests, review etc.).
- Open a separate PR where we can discuss whether to do more. Disseminating these types further isn't trivial (but still doable) and perhaps not even desired (because some consumers don't wish to distinguish and we don't want static/mirror details to leak into the more general API/codebase). Because of these uncertainties and their extent, I prefer not to insert a TODO in the code now and instead explore that path in a separate PR, where I can demonstrate concretely why it's perhaps dubious (much easier with a diff than in prose), if you'd like.

How does that sound?

File: pkg/kubelet/status/status_manager.go (hunk 1 above)
Comment (tallclair) [inline pkg/kubelet/status/status_manager.go]: SGTM.

**Comment by tallclair** [comment]
SGTM. Thanks!

**Comment by tallclair** [comment]
Please squash your commits after removing that comment, then I'll lgtm.

**Comment by ebbeelsborg** [comment]
@tallclair There's an e2e test failing, but I don't understand from the log details whether it's my fault or not. There is this (closed) issue: #43155.

I'm a bit surprised that these changes could have broken that test -- only the change to pkg/kubelet/status/BUILD looks possibly relevant in this regard to me. This is my first PR in kubernetes proper, so any hints are appreciated. Should I try to reproduce the error locally (following [](https://github.com/kubernetes/community/blob/master/contributors/devel/e2e-tests.md)) or do you know of a faster/better way to debug/fix/sidestep it? Thanks.

**Comment by ebbeelsborg** [comment]
Looking good, @tallclair. Who should assign @smarterclayton and ask for approval?

File: staging/src/k8s.io/apimachinery/pkg/types/uid.go [hunk 2]
```diff
@@ -20,3 +20,9 @@ package types
 // don't ONLY use UUIDs, this is an alias to string.  Being a type captures
 // intent and helps make sure that UIDs and names do not get conflated.
 type UID string
+
+// A pod UID which has been translated/resolved to the representation known to kubelets.
+type ResolvedPodUID UID
```
Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/types/uid.go]: If these are not accessed outside of the kubelet they don't belong in this package.  Would recommend kubelet types.

Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/types/uid.go]: Specifically `pkg/kubelet/types`, or a more specific one if these are not related.

Comment (ebbeelsborg) [inline staging/src/k8s.io/apimachinery/pkg/types/uid.go]: @smarterclayton Thanks, I've now done as you recommended. (They were placed there partly because they trivially use it as underlying type, and partly because the changes weren't contained to the kubelet package originally and might not be again after a envisioned second development iteration, but for this PR in isolation it belongs in the kubelet package.)

Comment (smarterclayton) [inline staging/src/k8s.io/apimachinery/pkg/types/uid.go]: We guard apimachinery/pkg/types very high.  It knows very little about anything other than core API concepts, of which the Kubelet and Pods do not overlap.  Core is where that distinction is made.
